File: utilities/bedutil-1.0.0/lib/regression.R

#######################################################################
# 'regression.R' takes a BED file from STDIN to which a second score
# has been appended as the 7th field.  A correlation plot is made and 
# a linear regression performed.  The original data are repeated to
# STDOUT and the regression data are printed to STDERR.
#----------------------------------------------------------------------
# Usage:
#     Rscript regression.R <jpgFile> <xLabel> <yLabel> <maxInputScore> <maxCalcScore>
# where the required parameters are:
#     jpgFile        name of jpg graph to be created
#     xLabel         label for the X-axis, i.e. the name of the input scores
#     yLabel         label for the Y-axis, i.e. the name of the appended scores
#     maxInputScore  maximum allowed value for the input scores
#                    any higher scores are set to maxInputScore prior to regression
#                    a value of zero enforces no limit
#     maxCalcScore   maximum allowed value for the appended (calculated) scores
#                    any higher scores are set to maxCalcScore prior to regression
#                    a value of zero enforces no limit
#######################################################################

# extract passed parameters
args <- commandArgs(TRUE)
jpgFile <- args[1]
xLabel <- args[2]
yLabel <- args[3]
maxInputScore <- as.numeric(args[4])
maxCalcScore <- as.numeric(args[5])

# collect and parse the input data
input <- read.table("stdin",header=FALSE,sep="\t")
inputScores <- input[,'V5']
calcScores <- input[,'V7']
if(maxInputScore != 0){
    inputScores <- sapply(inputScores, min, maxInputScore)
}
if(maxCalcScore != 0){
    calcScores <- sapply(calcScores, min, maxCalcScore) 
}
maxAdjInputScore <- max(inputScores)
maxAdjCalcScore <- max(calcScores)
solidPoints <- input[input$V5<=maxAdjInputScore & input$V7<=maxAdjCalcScore,c('V5','V7')]
openPoints  <- input[input$V5>maxAdjInputScore | input$V7>maxAdjCalcScore,c('V5','V7')]
if(maxInputScore != 0){
    openPoints$V5 <- sapply(openPoints$V5, min, maxInputScore)
}
if(maxCalcScore != 0){
    openPoints$V7 <- sapply(openPoints$V7, min, maxCalcScore)
}

# perform the linear regression
reg <- lm(calcScores~inputScores)

# create the visualization plot
width <- 2.5
height <- 2.5
units <- 'in' #w and h in inches
pointsize <- 7
res <- 600 #dpi
pch <- 20
cex <- 0.3
xlim <- c(0,maxAdjInputScore)
ylim <- c(0,maxAdjCalcScore)
bitmap(file=jpgFile,type='jpeg',width=width,height=height,unit=units,pointsize=pointsize,res=res)
plot(solidPoints[,'V5'],solidPoints[,'V7'],xlab=xLabel,ylab=yLabel,pch=20,xlim=xlim,ylim=ylim)
points(openPoints[,'V5'],openPoints[,'V7'],pch="+")
abline(reg=reg)

# print the results
write.table(input,file="",quote=FALSE,sep="\t",na="",row.names=FALSE,col.names=FALSE)
sink(stderr())
summary(reg)

File: utilities/bedutil-1.0.0/lib/regression.pl
#!/usr/bin/perl
use strict;
use warnings;
use Cwd qw(abs_path);

#######################################################################
# 'regression' takes feature BED lines on STDIN, streams them through 
# 'crossing' to collect a second score value, and then performs regression 
# analysis comparing the input scores (field 5) to the newly calculated scores.
# The correlation table is printed to STDOUT.  A plot of the simulation results
# is made and saved as a JPG image. 
#----------------------------------------------------------------------
# Options are:
#     JPG_FILE         name of the image file to create
#                      REQUIRED
#     X_LABEL          label for the X-axis, i.e. the name of the input values
#                      OPTIONAL [default: 'Input Value']
#     Y_LABEL          label for the Y-axis, i.e. the name of the calculated values
#                      OPTIONAL [default: 'Calculated Value']
#     MAX_INPUT_SCORE  maximum allowed value for the input scores
#                      any higher scores are set to $MAX_INPUT_SCORE prior to regression
#                      OPTIONAL [default: 0, no limit is enforced]
#     MAX_CALC_SCORE   label for the Y-axis, i.e. the name of the calculated values
#                      any higher scores are set to $MAX_CALC_SCORE prior to regression
#                      OPTIONAL [default: 0, no limit is enforced]
#----------------------------------------------------------------------
# To view the options inherited from crossing, use:
#     bedutil crossing --help
#----------------------------------------------------------------------
# Options may be provided as environment variables or on the command line in format:
#     <value_option>=<value> <boolean_option>  
# with no spaces, for example:
#     bedutil regression JPG_FILE=/path/to/my.jpg
# Options relevant to crossing must be passed as environment variables, 
# not on the command line.
#######################################################################


#######################################################################
# detect and check options
#----------------------------------------------------------------------
my $scriptFile = abs_path($0);
my ($scriptDir, $scriptName) = $scriptFile =~ m|(.+)/(.+)|;
require "$scriptDir/common.pl";
showHelp($scriptFile, 35, 29);
#----------------------------------------------------------------------
use vars qw(%validOptions %booleanOptions %defaultValues @requiredOptions);
%validOptions = map { $_ => 1} qw (
    JPG_FILE  
    X_LABEL
    Y_LABEL
    MAX_INPUT_SCORE
    MAX_CALC_SCORE
);
%booleanOptions = map { $_ => 1} qw (
);
%defaultValues = (
    X_LABEL => 'Input Value',
    Y_LABEL => 'Calculated Value',
    MAX_INPUT_SCORE => 0,
    MAX_CALC_SCORE => 0
);
@requiredOptions = qw (
    JPG_FILE
);
my ($error, $feedback) = parseOptions($scriptName);
#######################################################################


#######################################################################
# assemble and execute the stream
#----------------------------------------------------------------------
my $crossing =  getPerl("crossing");
my $regression =  "Rscript $scriptDir/regression.R ".
  "$ENV{JPG_FILE} \"$ENV{X_LABEL}\" \"$ENV{Y_LABEL}\" $ENV{MAX_INPUT_SCORE} $ENV{MAX_CALC_SCORE}";
my $stream = "$crossing | $regression";
open my $outH, "|-", $stream or die "regression.pl: could not open stream for writing: $!";
while(my $line = <STDIN>){
    print $outH $line;
}
close $outH;
#----------------------------------------------------------------------
sub getPerl {
    my ($function) = @_;
    return "perl $scriptDir/$function.pl";
}
#######################################################################

1;

